packages/@n8n/task-runner/src/js-task-runner/python-sandbox/python-sandbox.ts

```diff
@@ -0,0 +1,145 @@
+
+import type { SandboxContext } from './code-sandbox';
+import { CodeSandbox } from './code-sandbox';
+import { LoadPyodide } from './pyodide';
+import { freezePrototypes } from '../freeze-prototypes';
+
+type PythonSandboxContext = {
+	[K in keyof SandboxContext as K extends `$${infer I}` ? `_${I}` : K]: SandboxContext[K];
+};
+
+type PyodideError = Error & { type: string };
+
+const envAccessBlocked = process.env.N8N_BLOCK_ENV_ACCESS_IN_NODE === 'true';
+
+const PYTHON_HELPERS_CODE = `
+import js
+
+class Helpers:
+    def __init__(self):
+        self._js_helpers = _helpers_js
+    
+    async def httpRequest(self, options):
+        return await self._js_helpers.httpRequest(options)
+    
+    async def getBinaryDataBuffer(self, item_index, property_name):
+        result = await self._js_helpers.getBinaryDataBuffer(item_index, property_name)
+        return bytes(result.to_py()) # JS buffer to Python bytes
+    
+    async def prepareBinaryData(self, buffer, file_name=None, mime_type=None):
+        result = await self._js_helpers.prepareBinaryData(buffer, file_name, mime_type)
+        return result.to_py() # JS object to Python dict
+    
+    async def setBinaryDataBuffer(self, metadata, buffer):
+        result = await self._js_helpers.setBinaryDataBuffer(metadata, buffer)
+        return result.to_py() # JS object to Python dict
+    
+    async def binaryToString(self, buffer, encoding='utf8'):
+        result = await self._js_helpers.binaryToString(buffer, encoding)
+        return result # JS string (no conversion needed)
+    
+    async def assertBinaryData(self, item_index, property_name):
+        result = await self._js_helpers.assertBinaryData(item_index, property_name)
+        return result.to_py() # JS object to Python dict
+    
+    async def request(self, uri_or_object, options=None):
+        return await self._js_helpers.request(uri_or_object, options)
+
+helpers = Helpers()
+`;
+
+export class PythonSandbox extends CodeSandbox {
+	private readonly context: PythonSandboxContext;
+
+	constructor(
+		context: SandboxContext,
+		private pythonCode: string,
+		helpers: IExecuteFunctions['helpers'],
+		private readonly mode: 'secure' | 'insecure',
+	) {
+		super(helpers);
+		// Since python doesn't allow variable names starting with `$`,
+		// rename them to all to start with `_` instead
+		this.context = Object.keys(context).reduce((acc, key) => {
+			acc[key.startsWith('$') ? key.replace(/^\$/, '_') : key] = context[key];
+			return acc;
+		}, {} as PythonSandboxContext);
+	}
+
+	async runCode<T = unknown>(): Promise<T> {
+		return await this.runCodeInPython<T>();
+	}
+
+	async runCodeAllItems() {
+		return await this.runCodeInPython<INodeExecutionData[]>();
+	}
+
+	async runCodeEachItem(_itemIndex: number /* ignored as in Code node */) {
+		return await this.runCodeInPython<INodeExecutionData>();
+	}
+
+	private async runCodeInPython<T>() {
+		const packageCacheDir = await this.helpers.getStoragePath();
```

Comment:
The package dir should be an implementation detail of the python executor. AFAIK it doesn't need to be a helper. We can e.g. set it using env config or always default to some dir in `os.tmpdir` if not given.

It seems like `getStoragePath` was only added to provide a mechanism for a node to get a handle to some FS storage dir. Since this is now in the runner and not in the node, we don't need it anymore.